doc/pins/appl_state_db_response_path_hld.md

```diff
@@ -0,0 +1,218 @@
+
+* P4Orch does not retry on failed requests. \
+In the existing SONiC orch, failed requests will be queued and retried indefinitely until success. However, P4Orch does not retry any failed requests. The decision of retry can be done in the controller. If the orchagent does retry indefinitely, it should only send the response and update APPL STATE DB after the request is successfully programmed. In P4Orch, response can be sent immediately after noticing a success or failure since there is no retry.
+* P4Orch handles each P4RT request atomically. \
+From the controller perspective, each RPC is better to be an atomic operation. And hence P4Orch handles each P4RT request atomically. If there is any failure encountered during the attribute programming, P4Orch will do the best to revert the already programmed attributes. (If the reversion fails, P4Orch will notify the system with a critical error.) With this, P4Orch does not need to update the APPL STATE DB for a partially programmed request. It will only update APPL STATE DB after it fully programs a request.
+
+### APPL DB Cleanup
+
+Optionally, applications can choose to remove an intent from APPL DB if it failed to program into the hardware. This can be useful in warmboot since the device after warmboot should have the same hardware programming as before. Removing the failed entries from APPL DB can prevent those entries to be re-program after warmboot. Since the previous intent has failed, this cleanup operation in APPL DB is only at the DB level; orchagent does not need to process the cleanup operation. Applications should not use the ProducerStateTable to do the cleanup operation since it will trigger notifications to the orchagent to process the cleanup. Applications should directly modify APPL DB to revert the failed intent.
+
+The following are the types of failed intent and the possible cleanup operations:
+
+* Failed create \
+Delete the new key in APPL DB to cleanup.
+* Failed modify \
+Read APPL STATE DB for the original entry. Wipe the entry in APPL DB, and re-write it with the APPL STATE DB entry to cleanup.
+* Failed delete \
+Read APPL STATE DB for the original entry. Write it into APPL DB to cleanup.
+
+### PINS P4RT Examples
+
+#### Success Case
+
+![drawing](images/appl_state_db_response_path_hld_example_success.png)
+
+#### Failure Case (with APPL DB Cleanup)
+
+![drawing](images/appl_state_db_response_path_hld_example_failure.png)
+
+#### Batched Requests
+
+![drawing](images/appl_state_db_response_path_hld_example_batch.png)
+
+## SAI API
+
+N/A
+
+## Configuration and management
+
+A new configuration file will be added to enable the response channel and the APPL STATE DB for the selected APPL DB tables. There are two configurations: response channel and APPL STATE DB. They are defined in separate section of the configuration file. The configuration will contain a list of APPL DB table names. Only the tables in the list will have the feature enabled.
```

Comment:
Can you please list the location of this new configuration file?

Reply:
@donNewtonAlpha

We did not implement this approach.
We used a different approach to configure it in the request entry.
I have updated this section.

Reply:
Thanks for the update. It looks like some events in P4RT depend on updates from PORT_TABLE in APPL_STATE_DB. How will they now work?
A better question is how will non P4RT created objects get added to APPL_STATE_DB?

Reply:
In the MVP release, the response path support only applies to P4Orch and VrfOrch.
Non-p4rt objects are not in APPL STATE DB yet. They will need code change in each orch.

P4RT will wait for all ports ready before it starts.